widget/list.go

```diff
@@ -512,24 +316,96 @@ func (l *listLayout) MinSize(objects []fyne.CanvasObject) fyne.Size {
 	return fyne.NewSize(0, 0)
 }
 
-func (l *listLayout) getObjects() []fyne.CanvasObject {
-	objects := l.children
-	objects = append(objects, l.separators...)
-	return objects
+func (l *listLayout) getItem() *listItem {
+	item := l.itemPool.Obtain()
+	if item == nil {
+		if f := l.list.CreateItem; f != nil {
+			item = newListItem(f(), nil)
+		}
+	}
+	return item.(*listItem)
+}
+func (l *listLayout) offsetUpdated(pos fyne.Position) {
+	if l.list.offsetY == pos.Y {
+		return
+	}
+	l.list.offsetY = pos.Y
+	l.updateList()
 }
 
-func (l *listLayout) appendedItem(objects []fyne.CanvasObject) {
-	if len(objects) > 1 {
-		objects[len(objects)-1].Move(fyne.NewPos(0, objects[len(objects)-2].Position().Y+l.list.itemMin.Height+theme.SeparatorThicknessSize()))
-	} else {
-		objects[len(objects)-1].Move(fyne.NewPos(0, 0))
+func (l *listLayout) setupListItem(item fyne.CanvasObject, id ListItemID) {
+	li := item.(*listItem)
+	previousIndicator := li.selected
+	li.selected = false
+	for _, s := range l.list.selected {
+		if id == s {
+			li.selected = true
+		}
+	}
+	if previousIndicator != li.selected {
+		item.Refresh()
+	}
+	if f := l.list.UpdateItem; f != nil {
+		f(id, li.child)
+	}
+	li.onTapped = func() {
+		l.list.Select(id)
 	}
-	objects[len(objects)-1].Resize(fyne.NewSize(l.list.size.Width, l.list.itemMin.Height))
 }
 
-func (l *listLayout) prependedItem(objects []fyne.CanvasObject) {
-	objects[0].Move(fyne.NewPos(0, objects[1].Position().Y-l.list.itemMin.Height-theme.SeparatorThicknessSize()))
-	objects[0].Resize(fyne.NewSize(l.list.size.Width, l.list.itemMin.Height))
+func (l *listLayout) updateList() {
+	l.renderLock.Lock()
+	defer l.renderLock.Unlock()
+	separatorThickness := theme.SeparatorThicknessSize()
+	width := l.list.Size().Width
+	length := 0
+	if f := l.list.Length; f != nil {
+		length = f()
+	}
+	visibleItemCount := int(math.Ceil(float64(l.list.scroller.Size().Height)/float64(l.list.itemMin.Height+theme.SeparatorThicknessSize()))) + 1
+	offY := l.list.offsetY - float32(int(l.list.offsetY)%int(l.list.itemMin.Height+separatorThickness))
+	minRow := ListItemID(offY / (l.list.itemMin.Height + separatorThickness))
+	maxRow := ListItemID(fyne.Min(float32(minRow+visibleItemCount), float32(length)))
+
+	updateItem := l.list.UpdateItem
```

Comment:
I think you could just use this in the if statement below without assigning a variable for it.

Reply:
Good point, fixed